File: 0214-shortest-palindrome/0214-shortest-palindrome.py
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        # Reverse the string to check for overlaps with the prefix
        rev_s = s[::-1]
        # Combine the original string and reversed string with a special separator
        combined = s + "#" + rev_s
        
        # Compute the KMP table (partial match table)
        kmp_table = [0] * len(combined)
        
        for i in range(1, len(combined)):
            j = kmp_table[i - 1]
            while j > 0 and combined[i] != combined[j]:
                j = kmp_table[j - 1]
            if combined[i] == combined[j]:
                j += 1
            kmp_table[i] = j
        
        # The length of the palindrome is stored in the last position of the KMP table
        longest_palindrome_length = kmp_table[-1]
        
        # We need to add the reverse of the suffix (non-palindromic part) to the front
        suffix_to_add = rev_s[:len(s) - longest_palindrome_length]
        
        return suffix_to_add + s

File: 0214-shortest-palindrome/0214-shortest-palindrome.cpp
class Solution {
public:
    string shortestPalindrome(string s) {
        string rev_s = s;
        reverse(rev_s.begin(), rev_s.end());
        
        // Combine the original string, separator, and reversed string
        string combined = s + "#" + rev_s;
        int n = combined.size();
        
        // KMP partial match table
        vector<int> kmp_table(n, 0);
        
        // Build the KMP table
        for (int i = 1; i < n; ++i) {
            int j = kmp_table[i - 1];
            
            while (j > 0 && combined[i] != combined[j]) {
                j = kmp_table[j - 1];
            }
            if (combined[i] == combined[j]) {
                ++j;
            }
            kmp_table[i] = j;
        }
        
        // Find the length of the longest palindromic prefix
        int longest_palindrome_length = kmp_table[n - 1];
        
        // Add the necessary characters in front of the original string
        string suffix_to_add = rev_s.substr(0, s.size() - longest_palindrome_length);
        return suffix_to_add + s;
    }
};
